rpc/src/v1/impls/eth_filter.rs

```diff
@@ -143,127 +140,124 @@ impl<T: Filterable + Send + Sync + 'static> EthFilter for T {
+				let current_hashes = self.pending_transactions_hashes();
+
+				let new_hashes = {
+					// find all new hashes
+					current_hashes.difference(previous_hashes)
+						.cloned()
+						.map(Into::into)
+						.collect()
+				};
+
+				// save all hashes of pending transactions
+				*previous_hashes = current_hashes;
+
+				// return new hashes
+				Either::A(future::ok(FilterChanges::Hashes(new_hashes)))
+			},
+			PollFilter::Logs(ref mut block_number, ref mut last_block_hash, ref mut previous_logs, ref filter) => {
+				// retrive the current block number
+				let current_number = self.best_block_number();
+
+				// check if we need to check pending hashes
+				let include_pending = filter.to_block == Some(BlockNumber::Pending);
+
+				// build appropriate filter
+				let mut filter: EthcoreFilter = filter.clone().into();
+
+				// retrieve reorg logs
+				let (mut reorg, reorg_len) = last_block_hash.map_or_else(|| (Vec::new(), 0), |h| self.removed_logs(h, &filter));
+				*block_number -= reorg_len as u64;
+
+				filter.from_block = BlockId::Number(*block_number);
+				filter.to_block = BlockId::Latest;
+
+				// retrieve pending logs
+				let pending = if include_pending {
+					let pending_logs = self.pending_logs(current_number, &filter);
+
+					// remove logs about which client was already notified about
+					let new_pending_logs: Vec<_> = pending_logs.iter()
+						.filter(|p| !previous_logs.contains(p))
+						.cloned()
+						.collect();
+
+					// save all logs retrieved by client
+					*previous_logs = pending_logs.into_iter().collect();
+
+					new_pending_logs
+				} else {
+					Vec::new()
+				};
+
+				// save the number of the next block as a first block from which
+				// we want to get logs
+				*block_number = current_number + 1;
+
+				// save the current block hash, which we used to get back to the
+				// canon chain in case of reorg.
+				*last_block_hash = self.block_hash(BlockId::Number(current_number));
+
+				// retrieve logs in range from_block..min(BlockId::Latest..to_block)
+				let limit = filter.limit;
+				Either::B(self.logs(filter)
+					.map(move |logs| { reorg.extend(logs); reorg }) // append reorg logs in the front
+					.map(move |mut logs| { logs.extend(pending); logs }) // append fetched pending logs
+					.map(move |logs| limit_logs(logs, limit)) // limit the logs
+					.map(FilterChanges::Logs))
 			}
-		})
+		}))
 	}
 
 	fn filter_logs(&self, index: Index) -> BoxFuture<Vec<Log>> {
 		let filter = {
 			let mut polls = self.polls().lock();
 
-			match polls.poll(&index.value()) {
-				Some(&PollFilter::Logs(ref _block_number, ref _last_block_hash, ref _previous_log, ref filter)) => filter.clone(),
-				// just empty array
-				Some(_) => return Box::new(future::ok(Vec::new())),
+			match polls.poll(&index.value()).and_then(|f| f.modify(|filter| match *filter {
+				PollFilter::Logs(.., ref filter) => Some(filter.clone()),
+				_ => None,
+			})) {
+				Some(filter) => filter,
```

Comment:
Intentionally changed the behaviour here, returning empty vector when querying e.g. `pendingTransactionFilter` for logs was weird. I think "Filter not found" is more appropriate response.